platform/javascript/api/javascript_tools_editor_plugin.cpp

```diff
@@ -0,0 +1,123 @@
+#if defined(TOOLS_ENABLED) && defined(JAVASCRIPT_ENABLED)
+#include "javascript_tools_editor_plugin.h"
+
+#include "core/engine.h"
+#include "core/os/dir_access.h"
+#include "core/os/file_access.h"
+#include "core/project_settings.h"
+#include "editor/editor_node.h"
+
+#include <emscripten/emscripten.h>
+
+static void _javascript_editor_init_callback() {
+	EditorNode::get_singleton()->add_editor_plugin(memnew(JavaScriptToolsEditorPlugin(EditorNode::get_singleton())));
+}
+
+void JavaScriptToolsEditorPlugin::initialize() {
+	EditorNode::add_init_callback(_javascript_editor_init_callback);
+}
+
+JavaScriptToolsEditorPlugin::JavaScriptToolsEditorPlugin(EditorNode *p_editor) {
+	Variant v;
+	add_tool_menu_item("Download Project Source", this, "_download_zip", v);
+}
+
+void JavaScriptToolsEditorPlugin::_downalod_zip(Variant p_v) {
+	if (!Engine::get_singleton() || !Engine::get_singleton()->is_editor_hint()) {
+		WARN_PRINT("Project download is only available in Editor mode");
+		return;
+	}
+	String resource_path = ProjectSettings::get_singleton()->get_resource_path();
+
+	FileAccess *src_f;
+	zlib_filefunc_def io = zipio_create_io_from_file(&src_f);
+	zipFile zip = zipOpen2("/tmp/project.zip", APPEND_STATUS_CREATE, NULL, &io);
```

Comment:
I haven't tried, but make sure that it the generated temp files are properly cleaned both upon success and on failure.

Reply:
I do call `FS.unlink` in the JS code below, it should always work, the only case where it wouldn't is if `FS.open` or `FS.read` throws an exception (which is not handled), but in that case it probably means the file was not created anyway.
Should I add a try-catch? I don't think it's something that's recoverable anyway as it would mean some other underlying issue with the current runtime.

Reply:
IIUC this is the browser's own filesystem, right? I guess its `/tmp` might be cleaned when you close the browser so even if it were left in a broken state due to a bug, restarting the browser would fix it?

Reply:
Yeah, the emscripten filesystem is in-memory (stores a tree of names/arraybuffer), so closing the browser (or just the page) will clear it.
The only persistent part of the FS is `/userfs` which we mount as IDBFS. Plus, in the editor html, we also mount `/home/web_user/projects` and `/home/web_user/.config`